### PR TITLE
Resolving Issue 701: changing for-target-subscript to invalid-for-target

### LINKED ISSUE #701: for-target-subscript check should report attribute assignments used as loop target
Currently, `for-target-subscript` reports an error of the form `for x[0] in ...`, and we should extend this to `for x.attr in ...` as well.

Because this change makes the checker more general than just subscripts, we should do some renaming as well. As part of this task, rename the checker to `invalid-for-target` (files will need to be renamed as well).

### PR DESCRIPTION
**Pull request by ckatsube**
## Motivation and Context

resolves #701

## Your Changes

**Description**: changed the custom checker `InvalidForTargetChecker` (previously `ForTargetSubscriptChecker`) located in `python_ta/checkers/invalid_for_checker.py` (previously `for_target_subscript_checker.py`). This change entails updating the error message from `for-target-subscript` to `invalid-for-target` as well as changing the checker to add the message in the additional condition when a target of a for loop is an object's attribute (`for x.attr in ...`)

**Type of change** (select all that apply):

 

- [x] New feature (non-breaking change which adds functionality)

## Testing

Manually ran pylint on a test file using the CLI command `pylint --load-plugins="python_ta.checker.invalid_for_target_checker" sandbox.py` (where sandbox.py is an untracked local module containing the below code) and python_ta's `check_all` on the same test file.

```python
for x.attr in a:
    pass

for x[0] in a:
    pass
```

## Questions and Comments (if applicable)

There are a number of other references to the `for_target_subscript` checker. I was unsure of how many of them I should update the reference to in this pull request as the checker is fully functioning (these now-outdated references appear to be related to documentation and testing.
There are also a few references to `for-target-subscript`.
image
image
(Screenshot taken from PyCharm's "Find in Files" feature using the above mentioned search terms).

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the TravisCI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls.

### CODE REVIEW

**Comment by ckatsube** [comment]
@david-yz-liu 
I'm unsure of whether I should mark the PR as ready for review since I still have a few open questions. Would it be possible to get a brief review over the questions section to see if that's something I should continue working on before marking the PR as ready?

**Comment by ckatsube** [comment]
Just remembered another thing. When looking at `invalid-for-target`, I noticed that there is a check for if the for loop's target is a `tuple`, and to check if any of the assignments in said tuple are invalid targets. You can also explicitly write a list as a for loop target in something that would look like
```python
for [a, b] in l:
    pass
```
Should I add a check for lists into this PR as well? Or create a separate PR for it.

**Review by david-yz-liu — COMMENTED** [review]
@ckatsube thanks for the ping. Regarding your question, yes you should update *all* references to this checker, including in documentation and tests. In general it's common when updating code to update documentation and tests as well. In fact, you should create updated test cases for the changes you've made.

Also, yes please add a case for checking a `List` node as the for loop target (and tests as well). In fact, now that you pointed this out I'm thinking about cases where the for loop target is a nested expression with an invalid target in a subexpression, like

```python
for x, [y, z.attr] in lst:
```

So why don't you work on actually generalizing the checker by pulling out the actual logic that checks the target into a *recursive helper method*.

File: python_ta/checkers/invalid_for_target_checker.py [hunk 1]
```diff
@@ -6,32 +6,40 @@
...
+                if _target_is_subscript_or_assignattr(node):
+                    self.add_message('invalid-for-target',
                                      node=target, args=target.as_string())
 
 
+def _target_is_subscript_or_assignattr(node: astroid.For) -> bool:
+    target_is_subscript = isinstance(node.target, astroid.Subscript)
+    target_is_attr_assign = isinstance(node.target, astroid.AssignAttr)
+
+    return target_is_subscript or target_is_attr_assign
```
[the remark this answers is not in the document]
Comment (ckatsube) [reply python_ta/checkers/invalid_for_target_checker.py:41]: Noted! I'll get the above items completed before marking the PR as ready for review.

**Comment by ckatsube** [comment]
Brief Changelog:
- Generalized `InvalidForTargetChecker` in `invalid_for_target.py` to check for target lists as well as nested tuples/lists.
- Added `test_invalid_target_checker.py` file which contains a brief test suite for the `InvalidForTargetChecker`
- Updated references of the old message `for-target-subscript` to the new `invalid-for-target`

**Comment by ckatsube** [comment]
Unmarked as Ready for Review.
I realized that you can just use `.nodes_of_class` to extract all `Subscript` and `AssignAttr` nodes within the `For` node's target. Should we use this and check if these two lists are empty rather than flattening the target list and iterating through it?

**Comment by ckatsube** [comment]
Changelog:
Pretty much what the commit says: removed the helper that flattened the for node's target and instead used the `.nodes_of_class` node method to extract the invalid nodes from the target.

**Comment by ckatsube** [comment]
Changelog:
Pretty much what the commit says: removed the helper that flattened the for node's target and instead used the `.nodes_of_class` node method to extract the invalid nodes from the target.

**Comment by ckatsube** [comment]
Changelog:
Pretty much what the commit says: removed the helper that flattened the for node's target and instead used the `.nodes_of_class` node method to extract the invalid nodes from the target.

**Comment by ckatsube** [comment]
Changelog:
Pretty much what the commit says: removed the helper that flattened the for node's target and instead used the `.nodes_of_class` node method to extract the invalid nodes from the target.

File: python_ta/checkers/invalid_for_target_checker.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    name = 'invalid_for_target'
+    # use dashes for connecting words in message symbol
+    msgs = {'E9984': ('For loop variable "%s" should not be a part of a larger object.',
+                      'invalid-for-target',
+                      'Used when you have a loop variable in a for loop '
+                      'that is in subscript or object attribute form'),
+            }
+    # this is important so that your checker is executed before others
+    priority = -1
+    # pass in message symbol as a parameter of check_messages
```
Comment (david-yz-liu) [inline python_ta/checkers/invalid_for_target_checker.py:23]: You can delete this comment

File: python_ta/checkers/invalid_for_target_checker.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    INVALID_TARGETS = (astroid.Subscript, astroid.AssignAttr)
+
+    @check_messages('invalid-for-target')
+    def visit_for(self, node: astroid.For) -> None:
+
```
Comment (david-yz-liu) [inline python_ta/checkers/invalid_for_target_checker.py:29]: Delete this line

**Review by david-yz-liu — COMMENTED** [review]
@ckatsube great, I left a few minor comments that you should be able to resolve pretty quickly!

**Pull request merged**